---
title: "Cell competition: RNA bulk data from mouse cell lines BG (95%) and HB (24%)"
author: "Gabriele Lubatti"
date: "17/06/2021"
output: html_document
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
```

In this notebook it is shown the heteroplasmy analysis performed on bulk RNA seq mouse data from <https://www.biorxiv.org/content/10.1101/2020.01.15.900613v2>.
There are in total 8 samples that belong to the cell line BG (95%) and 8 samples that belongs to the cell line HB (24%). 
The two cell lines are built in order to have a different mitochondrial DNA from each other and from the reference genome in some bases.
The cell lines BG and HB are respectively classified as Loser and Winner. 

## Get counts for the four alleles in each base-sample pair

```{r}

#if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
#devtools::install_github("ScialdoneLab/RNAheteroplasmy",auth_token="be68e5a7d8bdb8c1d270488c463c6bcaa5#13acf7")
#library(RNAheteroplasmy)
```


```{r}
your_path="/home/ies/gabriele.lubatti/revision_heteroplasmy"
```

```{r}


setwd(paste0(your_path,"/RNAHeteroplasmynew-main/Packages_Functions"))
source("RNAheteroplasmynew_packages.R")
source("get_heteroplasmy.R")
source("get_raw_counts_allele.R")
source("lineage_tracing.R")
source("other_functions.R")
source("plotting_functions.R")
source("statistical_test.R")

```

```{r}
#if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
#devtools::install_github("https://github.com/ScialdoneLab/RNAheteroplasmy_new/tree/master",auth_token=#"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")
#library(RNAheteroplasmynew)
```

The first step of the library RNAheteroplasmy is to generate a raw counts allele matrix with cells as rows and the four alleles for each base in the fasta file on the columns.
This task is achieved with the function *get_raw_counts_allele*. As input we need to provide the sorted bam files (one for each cell, with full path), the fasta file of the genomic region of interested and the cell names.
The matrixes *meta_data_ana_final_big* and *meta_data_ana_final_small*  contain meta data information about the cells (i.e. cell names, condition). 

```{r}
#setwd("/Users/gabriele.lubatti/Desktop/Phd/Heteroplasmy_library_R/Input_data_R_server")
#load("output_SNP_ana_mt.Rda")
#data("output_SNP_ana_mt",package="RNAheteroplasmy")


```

```{r}

load(paste0(your_path,"/RNAHeteroplasmynew-master/data/","meta_data_ana_final_big.rda"))
load(paste0(your_path,"/RNAHeteroplasmynew-master/data/","meta_data_ana_final_small.rda"))
#data("meta_data_ana_final_big",package="RNAheteroplasmynew")
#data("meta_data_ana_final_small",package="RNAheteroplasmynew")


path_to_bam="/home/ies/gabriele.lubatti/revision_heteroplasmy/Ana_data/bam_file_ana_sorted/"
cell_names=as.vector(meta_data_ana_final_big$name_cell)

bam_input=paste(path_to_bam,cell_names,".unique.bam",sep="")

path_fasta="/home/ies/gabriele.lubatti/revision_heteroplasmy/Cell_Competition_data/fasta_files/Mus_musculus.GRCm38.dna.chromosome.MT.fa"
cell_names_bulk=cell_names
bam_input_bulk=bam_input
#setwd("/Users/gabriele.lubatti/Desktop/Phd/Heteroplasmy_library_R/Input_data_R_server")
#save(cell_names_bulk,file="cell_names_bulk.Rda")
#save(bam_input_bulk,file="bam_input_bulk.Rda")

```

We don't execute the function *get_raw_counts_allele* here and  we directly load his output. 
A command line implentation of the function *get_raw_counts_allele* is also available (see github README file for info).

```{r}
#output_SNP_ana_mt=get_raw_counts_allele(bam_input_bulk,path_fasta,cell_names_bulk)

# Command line implementation
#cd /home/ies/gabriele.lubatti/revision_heteroplasmy/Cell_Competition_data/R_files
#Rscript --vanilla /home/ies/gabriele.lubatti/revision_heteroplasmy/Cell_Competition_data/R_files/get_r#aw_counts_allele_script.R -b bam_input_bulk -f #"/home/ies/gabriele.lubatti/revision_heteroplasmy/Cell_Competition_data/fasta_files/Mus_musculus.GRCm3#8.dna.chromosome.MT.fa" -c cell_names_bulk -o output_SNP_ana_mt.Rda  -s 20

#data("output_SNP_ana_mt",package="RNAheteroplasmynew")

load(paste0(your_path,"/RNAHeteroplasmynew-master/data/","output_SNP_ana_mt.rda"))
```


The output of *get_raw_counts_allele* is a list with three elements (see *?get_raw_counts_allele* for more info). The first element is the matrix of counts (n_rows = number of cells, n_cols= 4*number of bases) of the four alleles in each base. The row names are equal to cell_names.

```{r}
#output_SNP_ana_mt=result

```



```{r}
matrix_allele_counts=output_SNP_ana_mt[[1]]
name_position_allele=output_SNP_ana_mt[[2]]
name_position=output_SNP_ana_mt[[3]]

```

```{r}
row.names(meta_data_ana_final_big)=meta_data_ana_final_big$name_cell
meta_data_ana_final_big=meta_data_ana_final_big[row.names(matrix_allele_counts),]
delete_duplicate=meta_data_ana_final_big$name_cell[seq(1,48,3)]
meta_data_ana_final_big_filter=meta_data_ana_final_big[delete_duplicate,]
bulk_sample=meta_data_ana_final_big_filter$name_cell_original

matrix_allele_counts=matrix_allele_counts[delete_duplicate,]
row.names(matrix_allele_counts)=bulk_sample


row.names(meta_data_ana_final_small)=meta_data_ana_final_small$name_fastq
```


The next step is to obtain a matrix with allele frequencies and a matrix with heteroplasmy values for each pair of cell-base. This is obtained with the function *get_heteroplasmy*. 
This function performs a two step filtering procedure, the first on the cells and the second on the bases. The aim is to keep only the cells that have more than *number_reads* counts in more than *number_positions* bases and to keep only the bases that are covered by more than *number_reads* counts in all the cells (*filtering*=1)  or in at least 50% of cells in each cluster (*filtering*=2).


```{r}

bulk_data_competition=get_heteroplasmy(matrix_allele_counts[bulk_sample,],name_position_allele,name_position,50,2000,filtering = 1)

```

Among the ouptut of *get_geteroplasmy* there are the matrix with heteroplasmy values and the matrix with allele frequencies, for all the cells and bases that pass the two step filtering procedure. 
The heteroplasmy is computed as *1-max(f)*, where *f* are the frequencies of the four alleles for every cell-base pair.
For more info about the ouput see *?get_geteroplasmy*.

```{r}
sum_matrix=bulk_data_competition[[1]]
sum_matrix_qc=bulk_data_competition[[2]]
heteroplasmy_matrix_bulk=bulk_data_competition[[3]]
allele_matrix_bulk=bulk_data_competition[[4]]
cluster_bulk=as.character(meta_data_ana_final_small[row.names(heteroplasmy_matrix_bulk),]$status)
condition_bulk=as.character(meta_data_ana_final_small[row.names(heteroplasmy_matrix_bulk),]$condition)
index_bulk=bulk_data_competition[[5]]
```




```{r}

name_position_allele_qc=name_position_allele[name_position%in%colnames(sum_matrix_qc)]
name_position_qc=name_position[name_position%in%colnames(sum_matrix_qc)]

```

It is possible to perform an additional filtering step on the bases keeping only the ones with an heteroplasmy value above *min_heteroplasmy* in more than *min_cells*.

```{r}
relevant_bases=filter_bases(heteroplasmy_matrix_bulk,0.01,8)
```


We can compute and visualize the distribution of the average coverage of the bases.

```{r}

max_entropy_base=get_distribution(heteroplasmy_matrix_bulk[,relevant_bases],"max")
mean_counts_base=get_distribution(sum_matrix_qc[,relevant_bases],"mean")

```

```{r}
Plot_distribution(mean_counts_base,"mean counts/base","Distribution mean counts/base")

```
## Identification of most different bases according to heteroplasmy between clusters
For detecting the difference in heteroplasmy values between two group of cells (i.e. two clusters), an unpaired two-samples Wilcoxon test is performed. In this case we run the test between the groups *Winner* and *Loser*. As output, for each bases, there is the adjust p value (FDR).

```{r}
p_value_wilcox_test=get_wilcox_test(heteroplasmy_matrix_bulk[,relevant_bases],cluster_bulk,"Loser","Winner" )
```

We sort the bases according to the adjust p value in order to identify the bases where the heteroplasmy is most different between the two groups.
```{r}
p_value_wilcox_test_sort=sort(p_value_wilcox_test,decreasing = F)

```

The heteroplasmy and the corresponding allele frequencies for the most relevant bases (according to Wilcoxon test) are shown. The difference in heteroplasmy values is sharp, as expected since the two cell lineages (Winner and Loser) have a different mitochondrial DNA by construction.

```{r}


q=list()
for ( i in 1:length(p_value_wilcox_test_sort[1:2])){
p=plot_heteroplasmy(names(p_value_wilcox_test_sort)[i],heteroplasmy_matrix_bulk,cluster_bulk)+ggtitle(paste(names(p_value_wilcox_test_sort)[i],round(p_value_wilcox_test_sort[i],4),sep="-"))
q=list(q,p)
}
q



#q=list()
#for ( i in names(p_value_wilcox_test_sort)[1:2]){
#p=plot_allele_frequency(i,heteroplasmy_matrix_bulk,allele_matrix_bulk,cluster_bulk,name_position_qc,na#me_position_allele_qc,5,title = "Allele frequencies")  
#q=list(q,p)
#}
#q

q=list()
for ( i in names(p_value_wilcox_test_sort)[1:2]){
p=plot_allele_frequency(i,heteroplasmy_matrix_bulk,allele_matrix_bulk,cluster_bulk,name_position_qc,name_position_allele_qc,5)  
q=list(q,p)
}
q


                                                            
```

We can check if there is a batch effect in the most relevan positions (i.e. the heteroplasmy levels are constantly higher only in a specific batch). For the top 2 positions defined with the GAM fit there is not a batch effect, since the difference in  heteroplasmy levels are driven by group (Winner and Loser) and not by batch (co culture-CO or separate culture-Sep).


```{r}
utile=meta_data_ana_final_small[row.names(heteroplasmy_matrix_bulk),]
batch=rep(0,length(utile$condition))
batch[utile$status=="Loser"&utile$condition=="CO"]="Loser-CO"
batch[utile$status=="Loser"&utile$condition=="Sep"]="Loser-Sep"
batch[utile$status=="Winner"&utile$condition=="CO"]="Winner-CO"
batch[utile$status=="Winner"&utile$condition=="Sep"]="Winner-Sep"




colour=rep(0,length(batch))
colour[grep("Loser",batch)]=gg_color_hue(2)[1]
colour[grep("Winner",batch)]=gg_color_hue(2)[2]

cluster=utile$status

q=list()
for ( i in names(p_value_wilcox_test_sort)[1:2]){
p=plot_batch(i,heteroplasmy_matrix_bulk,batch,colour,cluster,6)
q=list(q,p)
}
q


```

## Cluster analysis among samples based on allele frequency values
RNAheteroplasmy offers the possibility to perform a hierarchical clustering on the samples based on a distance matrix with the function *clustering_dist_ang*. Given a base, the distance between two sample is the angular distance of the allele frequencies. In this way we can represent the difference between two samples as a vector whose cordinates are the angular distances of the bases.
The total distance between two sample is the euclidean norm of the vector of difference between the two samples.
The output of *clustering_dist_ang* is a list with two elements. The first is a data frame which contains the old classification (partion available before the cluster analysis based on allele frequencies) and the new classification (partion provided by the cluster analysis based on allele frequencies ). The second is the distance matrix, on which the hierarchical clustering is done.
The heatmap of the distance matrix with samples sorted according to the new classification is shown below.
See *?clustering_dist_ang* for more info.





```{r}
result_clustering_bulk=clustering_dist_ang(heteroplasmy_matrix_bulk,allele_matrix_bulk,cluster_bulk,100,deepSplit_param=0,minClusterSize_param=2)
old_new_classification=result_clustering_bulk[[1]]
dist_matrix_bulk=result_clustering_bulk[[2]]

old_classification=as.vector(old_new_classification[,1])
new_classification=as.vector(old_new_classification[,2])

heatmap_plot(row.names(dist_matrix_bulk),row.names(dist_matrix_bulk),new_classification,old_classification,as.matrix(dist_matrix_bulk),cluster_columns=T,cluster_rows=T,"Euclidean distance")
```


The new classification coincide with the old classification. This means that we can perfectly distinguish between the two cell lineages only by looking at the heteroplasmy values of the mitochondrial bases. This result was expected since the samples derived from two cell lineage, whose mitochondrial DNA is built to be different between each other.

## Heteroplasmy and allele frequencies analysis for bases different from BL6 reference genome

As anticipated at the beginning of the notebook, the two cell lines are built in order to have a different mitochondrial DNA than the reference genome in some bases. Since the mt DNA sequence of the two cell lines is available (<https://www.ncbi.nlm.nih.gov/nuccore/KC663619.1>  and <https://www.ncbi.nlm.nih.gov/nuccore/KC663620.1>), we can identify the bases that are different from the reference genome
```{r}

bg_path=paste0(your_path,"/RNAHeteroplasmynew-master/inst/extdata/","bg.fasta.txt")
#bg_path=system.file("extdata", "bg.fasta.txt", package = "RNAheteroplasmynew")

fastaFile <- readDNAStringSet(bg_path)
seq_name = names(fastaFile)
sequence = paste(fastaFile)
df_bg <- data.frame(seq_name, sequence)

bg_sequence=df_bg$sequence

hb_path=paste0(your_path,"/RNAHeteroplasmynew-master/inst/extdata/","hb.fasta.txt")
#hb_path=system.file("extdata", "hb.fasta.txt", package = "RNAheteroplasmynew")
fastaFile <- readDNAStringSet(hb_path)
seq_name = names(fastaFile)
sequence = paste(fastaFile)
df_hb <- data.frame(seq_name, sequence)

hb_sequence=df_hb$sequence

#mt_path=system.file("extdata", "Mus_musculus.GRCm38.dna.chromosome.MT.fa", package = #"RNAheteroplasmynew")

mt_path=paste0(your_path,"/RNAHeteroplasmynew-master/inst/extdata/","Mus_musculus.GRCm38.dna.chromosome.MT.fa")

fastaFile <- readDNAStringSet(mt_path)
seq_name = names(fastaFile)
sequence = paste(fastaFile)
df_ref <- data.frame(seq_name, sequence)

ref_sequence=df_ref$sequence


hb_sequence=unlist(strsplit(hb_sequence,split=NULL))
ref_sequence=unlist(strsplit(ref_sequence,split=NULL))
bg_sequence=unlist(strsplit(bg_sequence,split=NULL))

```

The mt DNA sequences for the two cell lines have some insertions, since they are longer than the reference. We detect and remove the insertions with the function *detect_insertion* in order to have the two mt DNA sequences (from the two cell line) of the same length as the reference.
See *?detect_insertion* for more info.





```{r }

hb_sequence=detect_insertion(ref_sequence,hb_sequence)
bg_sequence=detect_insertion(ref_sequence,bg_sequence)
```

```{r }
position_kept=substr(colnames(heteroplasmy_matrix_bulk),1,nchar(colnames(heteroplasmy_matrix_bulk))-3)
position_kept=as.numeric(position_kept)
```

### Loser cell line (BG line-95%)
The heteroplasmy level (average among all the samples) is shown for all the bases, divided between the ones that are different (*Different Position*) and equal (*Equal Position*) with respect to the reference. 
```{r }
pos_diff_bg=bg_sequence[which(ref_sequence!=bg_sequence)]
index_diff_bg=which(ref_sequence!=bg_sequence)
pos_diff_bg=pos_diff_bg[index_diff_bg%in%position_kept]
```


```{r}
index_loser=which(cluster_bulk=="Loser")
heteroplasmy_matrix_bulk_bg=heteroplasmy_matrix_bulk[index_loser,]
allele_matrix_bulk_bg=allele_matrix_bulk[index_loser,]
cluster_bulk_bg=cluster_bulk[index_loser]
mean_heteroplasmy=apply(heteroplasmy_matrix_bulk_bg,2,median)
mean_heteroplasmy_diff=mean_heteroplasmy[position_kept%in%index_diff_bg]
mean_heteroplasmy_equal=mean_heteroplasmy[!position_kept%in%index_diff_bg]
diff_pos=rep('Different Position',length(mean_heteroplasmy_diff))
equal_pos=rep('Equal Position',length(mean_heteroplasmy_equal))
cond_pos=c(diff_pos,equal_pos)
mean_pos=c(mean_heteroplasmy_diff,mean_heteroplasmy_equal)

```




```{r}
Plot_boxplot(mean_heteroplasmy_diff,mean_heteroplasmy_equal,diff_pos,equal_pos,"Condition Position","Mean Heteroplasmy","BG_strain")

```

The values of heteroplasmy we found in the *Different Positions* with our computational analysis were very close to those estimated by ARMS-qPCR: for BG(95%), ~7% from RNA-seq data vs ~5% measured by ARMS-qPCR (see <https://www.biorxiv.org/content/10.1101/2020.01.15.900613v2> for more details on ARMS-qPCR).

We then check, for the *Different Positions*, if the dominant allele coincide with the reference allele in BG line, as expected. This is indeed the case for the vast majority of the *Different Positions*.





```{r}
position_name_diff_bg=colnames(heteroplasmy_matrix_bulk_bg)[position_kept%in%index_diff_bg]

#q=list()
#for ( i in 1:length(position_name_diff_bg[1:2])){
#p=plot_allele_frequency(position_name_diff_bg[i],heteroplasmy_matrix_bulk_bg,allele_matrix_bulk_bg,clu#ster_bulk_bg,name_position_qc,name_position_allele_qc,5,paste("Reference allele in the #strain:",pos_diff_bg[i]))
#q=list(q,p)
#}
#q


q=list()
for ( i in 1:length(position_name_diff_bg[1:2])){
p=plot_allele_frequency(position_name_diff_bg[i],heteroplasmy_matrix_bulk_bg,allele_matrix_bulk_bg,cluster_bulk_bg,name_position_qc,name_position_allele_qc,5)
q=list(q,p)
}
q

for ( i in 1:length(position_name_diff_bg[1:2])){
print(paste("Reference allele in the strain:",pos_diff_bg[i]))}

```

### Winner cell line (HB line-24%)
The heteroplasmy level (average among all the samples) is shown for all the bases, divided between the ones that are different (*Different Position*) and equal (*Equal Position*) with respect to the reference.
```{r }
pos_diff_hb=hb_sequence[which(ref_sequence!=hb_sequence)]
index_diff_hb=which(ref_sequence!=hb_sequence)
pos_diff_hb=pos_diff_hb[index_diff_hb%in%position_kept]
```


```{r}
index_winner=which(cluster_bulk=="Winner")
heteroplasmy_matrix_bulk_hb=heteroplasmy_matrix_bulk[index_winner,]
allele_matrix_bulk_hb=allele_matrix_bulk[index_winner,]
cluster_bulk_hb=cluster_bulk[index_winner]
mean_heteroplasmy=apply(heteroplasmy_matrix_bulk_hb,2,median)
mean_heteroplasmy_diff=mean_heteroplasmy[position_kept%in%index_diff_hb]
mean_heteroplasmy_equal=mean_heteroplasmy[!position_kept%in%index_diff_hb]
diff_pos=rep('Different Position',length(mean_heteroplasmy_diff))
equal_pos=rep('Equal Position',length(mean_heteroplasmy_equal))
cond_pos=c(diff_pos,equal_pos)
mean_pos=c(mean_heteroplasmy_diff,mean_heteroplasmy_equal)

```


```{r}
Plot_boxplot(mean_heteroplasmy_diff,mean_heteroplasmy_equal,diff_pos,equal_pos,"Condition Position","Mean Heteroplasmy","HB_strain")

```


The values of heteroplasmy we found in the *Different Positions* with our computational analysis were very close to those estimated by ARMS-qPCR: for HB(24%), ~17% from RNA-seq data vs ~24% measured by ARMS-qPCR (see <https://www.biorxiv.org/content/10.1101/2020.01.15.900613v2> for more details on ARMS-qPCR).

We then check, for the *Different Positions*, if the second dominant allele coincide with the reference allele in HB line, as expected. This is indeed the case for the vast majority of the *Different Positions*.

```{r}


position_name_diff_hb=colnames(heteroplasmy_matrix_bulk_hb)[position_kept%in%index_diff_hb]

#q=list()
#for ( i in 1:length(position_name_diff_hb[1:2])){
#p=plot_allele_frequency(position_name_diff_hb[i],heteroplasmy_matrix_bulk_hb,allele_matrix_bulk_hb,clu#ster_bulk_hb,name_position_qc,name_position_allele_qc,5,paste("Reference allele in the #strain:",pos_diff_hb[i]))
#q=list(q,p)
#}
#q


q=list()
for ( i in 1:length(position_name_diff_hb[1:2])){
p=plot_allele_frequency(position_name_diff_hb[i],heteroplasmy_matrix_bulk_hb,allele_matrix_bulk_hb,cluster_bulk_hb,name_position_qc,name_position_allele_qc,5)
q=list(q,p)
}
q

for ( i in 1:length(position_name_diff_hb[1:2])){
print(paste("Reference allele in the strain:",pos_diff_hb[i]))
}

```


Finally we observe that the 98% of the top 50 bases defined in a unsupervised way (Wilcoxon test) are also bases that are different from the BL6 reference genome in the two cell lines. This reveals the high consistency of the analysis performed by RNAheteroplasmy. 
```{r}
all_position_name_diff=c(position_name_diff_bg,position_name_diff_hb)
sum(unique(all_position_name_diff)%in%names(p_value_wilcox_test_sort)[1:50])/50
```



```{r}
library(utils)
utils::sessionInfo()
```





