<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cell_competition_mt_example_notebook • MitoHEAR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cell_competition_mt_example_notebook">
<meta property="og:description" content="MitoHEAR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MitoHEAR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Ludwig_et_al_example_notebook.html">Ludwig_et_al_example_notebook</a>
    </li>
    <li>
      <a href="../articles/cell_competition_bulk_data_mt_example_notebook.html">cell_competition_bulk_data_mt_example_notebook</a>
    </li>
    <li>
      <a href="../articles/cell_competition_mt_example_notebook.html">cell_competition_mt_example_notebook</a>
    </li>
    <li>
      <a href="../articles/lineage_tracing_example_notebook.html">lineage_tracing_example_notebook</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ScialdoneLab/MitoHEAR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>cell_competition_mt_example_notebook</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ScialdoneLab/MitoHEAR/blob/HEAD/vignettes/cell_competition_mt_example_notebook.Rmd" class="external-link"><code>vignettes/cell_competition_mt_example_notebook.Rmd</code></a></small>
      <div class="hidden name"><code>cell_competition_mt_example_notebook.Rmd</code></div>

    </div>

    
    
<p>The notebook depends on karyoploteR, regioneR, gam and fmsb packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MitoHEAR</span><span class="op">)</span>
<span class="va">required</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"karyoploteR"</span>, <span class="st">"regioneR"</span>, <span class="st">"gam"</span>, <span class="st">"fmsb"</span><span class="op">)</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">required</span>, <span class="kw">function</span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="va">pkg</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>eval <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>In this notebook it is shown the heteroplasmy analysis performed on mouse CI-treated epiblast cells from <a href="https://www.nature.com/articles/s42255-021-00422-7" class="external-link">Lima et al, 2021</a>.</p>
<div class="section level3">
<h3 id="get-counts-for-the-four-alleles-in-each-base-cell-pair">Get counts for the four alleles in each base-cell pair<a class="anchor" aria-label="anchor" href="#get-counts-for-the-four-alleles-in-each-base-cell-pair"></a>
</h3>
<p>The first step of the library MitoHEAR is to generate a raw counts allele matrix with cells as rows and the four alleles for each base in the fasta file on the columns. This task is achieved with the function <em>get_raw_counts_allele</em>. As input we need to provide the sorted bam files (one for each cell, with full path), the fasta file of the genomic region of interested and the cell names. The matrix <em>after_qc</em> contains meta data information about the cells (i.e. cell names, cell types, batch).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"after_qc.Rda"</span>, package <span class="op">=</span> <span class="st">"MitoHEAR"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cell_names</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">after_qc</span><span class="op">$</span><span class="va">new_name</span><span class="op">)</span></code></pre></div>
<p>We don’t execute the function <em>get_raw_counts_allele</em> here and we directly load his output. A command line implementation of the function <em>get_raw_counts_allele</em> is also available (see github README file for info).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"output_SNP_mt.Rda"</span>, package <span class="op">=</span> <span class="st">"MitoHEAR"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The output of <em>get_raw_counts_allele</em> is a list with three elements (see <em>?get_raw_counts_allele</em> for more info). The first element is the matrix of counts (n_rows = number of cells, n_cols= 4*number of bases) of the four alleles in each base. The row names are equal to cell_names.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">matrix_allele_counts</span> <span class="op">&lt;-</span> <span class="va">output_SNP_mt</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">name_position_allele</span> <span class="op">&lt;-</span> <span class="va">output_SNP_mt</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">name_position</span> <span class="op">&lt;-</span> <span class="va">output_SNP_mt</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>We select only the CI-treated epiblast cells for the down-stream analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">after_qc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">$</span><span class="va">new_name</span>
<span class="va">cells_fmk_epi</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">[</span><span class="op">(</span><span class="va">after_qc</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"Cell competition OFF"</span><span class="op">)</span><span class="op">&amp;</span><span class="op">(</span><span class="va">after_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">|</span><span class="va">after_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">|</span><span class="va">after_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span>, <span class="st">"new_name"</span><span class="op">]</span>

<span class="va">after_qc_fmk_epi</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">[</span><span class="va">cells_fmk_epi</span>, <span class="op">]</span>
<span class="va">my.clusters</span> <span class="op">&lt;-</span> <span class="va">after_qc_fmk_epi</span><span class="op">$</span><span class="va">cluster</span></code></pre></div>
<p>The next step is to obtain a matrix with allele frequencies and a matrix with heteroplasmy values for each pair of cell-base. This is obtained with the function <em>get_heteroplasmy</em>. This function performs a two step filtering procedure, the first on the cells and the second on the bases. The aim is to keep only the cells that have more than <em>number_reads</em> counts in more than <em>number_positions</em> bases and to keep only the bases that are covered by more than <em>number_reads</em> counts in all the cells (<em>filtering</em>=1) or in at least 50% of cells in each cluster (<em>filtering</em>=2).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="va">epiblast_ci</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_heteroplasmy.html">get_heteroplasmy</a></span><span class="op">(</span><span class="va">matrix_allele_counts</span><span class="op">[</span><span class="va">cells_fmk_epi</span>, <span class="op">]</span>, <span class="va">name_position_allele</span>, <span class="va">name_position</span>, number_reads <span class="op">=</span> <span class="fl">50</span>, number_positions <span class="op">=</span> <span class="fl">2000</span>, filtering <span class="op">=</span> <span class="fl">2</span>, <span class="va">my.clusters</span><span class="op">)</span></code></pre></div>
<p>Among the output of <em>get_heteroplasmy</em> there are the matrix with heteroplasmy values and the matrix with allele frequencies, for all the cells and bases that pass the two step filtering procedure. The heteroplasmy is computed as <em>1-max(f)</em>, where <em>f</em> are the frequencies of the four alleles for every cell-base pair. For more info about the output see <em>?get_heteroplasmy</em>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sum_matrix</span> <span class="op">&lt;-</span> <span class="va">epiblast_ci</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">sum_matrix_qc</span> <span class="op">&lt;-</span> <span class="va">epiblast_ci</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">heteroplasmy_matrix_ci</span> <span class="op">&lt;-</span> <span class="va">epiblast_ci</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="va">allele_matrix_ci</span> <span class="op">&lt;-</span> <span class="va">epiblast_ci</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>
<span class="va">cluster_ci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">after_qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>
<span class="va">cluster_ci</span><span class="op">[</span><span class="va">cluster_ci</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Winner Epiblast"</span>
<span class="va">cluster_ci</span><span class="op">[</span><span class="va">cluster_ci</span> <span class="op">==</span> <span class="st">"3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Intermediate"</span>
<span class="va">cluster_ci</span><span class="op">[</span><span class="va">cluster_ci</span> <span class="op">==</span> <span class="st">"4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Loser Epiblast"</span>
<span class="va">condition_ci</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">after_qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span>
<span class="va">index_ci</span> <span class="op">&lt;-</span> <span class="va">epiblast_ci</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>Below the customized function <em>plot_cells_coverage_epiblast</em> is used in order to keep the same colour assignment and notation used in <a href="https://www.nature.com/articles/s42255-021-00422-7" class="external-link">Lima et al, 2021</a>. For general use, please use function <em>plot_cells_coverage</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_cells_coverage_epiblast</span><span class="op">(</span><span class="va">sum_matrix_qc</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sum_matrix_qc</span><span class="op">)</span>, <span class="va">cluster_ci</span>, interactive<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">



<span class="fu"><a href="../reference/plot_base_coverage.html">plot_base_coverage</a></span><span class="op">(</span><span class="va">sum_matrix</span>, <span class="va">sum_matrix_qc</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sum_matrix_qc</span><span class="op">)</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span>, text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>We can visualize the mitochondrial genome showing in red the positions that passed our filtering based on coverage and were considered for the heteroplasmy analysis. The plot is based on the R package <em>karyoploteR</em>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">chr_name</span> <span class="op">&lt;-</span> <span class="st">"MT"</span>
<span class="va">heteroplasmy_matrix</span> <span class="op">&lt;-</span> <span class="va">heteroplasmy_matrix_ci</span>
<span class="va">biomart_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mt_start_end.txt"</span>, package <span class="op">=</span> <span class="st">"MitoHEAR"</span><span class="op">)</span>
<span class="va">biomart_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">biomart_file</span>, sep <span class="op">=</span>  <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">fastaFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Mus_musculus.GRCm38.dna.chromosome.MT.fa"</span>, package <span class="op">=</span> <span class="st">"MitoHEAR"</span><span class="op">)</span>
<span class="va">fastaFile</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">fastaFile</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_genome_coverage.html">plot_genome_coverage</a></span><span class="op">(</span><span class="va">biomart_file</span>, <span class="va">fastaFile</span>, <span class="va">chr_name</span>, <span class="va">heteroplasmy_matrix</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">name_position_allele_qc</span> <span class="op">&lt;-</span> <span class="va">name_position_allele</span><span class="op">[</span><span class="va">name_position</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sum_matrix_qc</span><span class="op">)</span><span class="op">]</span>
<span class="va">name_position_qc</span> <span class="op">&lt;-</span>  <span class="va">name_position</span><span class="op">[</span><span class="va">name_position</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sum_matrix_qc</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>It is possible to perform an additional filtering step on the bases keeping only the ones with an heteroplasmy value above <em>min_heteroplasmy</em> in more than <em>min_cells</em>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">relevant_bases</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_bases.html">filter_bases</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span>, min_heteroplasmy <span class="op">=</span> <span class="fl">0.01</span>, min_cells <span class="op">=</span> <span class="fl">10</span>, <span class="va">index_ci</span><span class="op">)</span></code></pre></div>
<p>We can compute and visualize the distribution of the average coverage of the bases.</p>
</div>
<div class="section level3">
<h3 id="identification-of-most-different-bases-according-to-heteroplasmy-between-clusters">Identification of most different bases according to heteroplasmy between clusters<a class="anchor" aria-label="anchor" href="#identification-of-most-different-bases-according-to-heteroplasmy-between-clusters"></a>
</h3>
<p>For detecting the difference in heteroplasmy values between two group of cells (i.e. two clusters), an unpaired two-samples Wilcoxon test is performed. In this case we run the test between the clusters <em>Winner Epiblast</em> and <em>Loser Epiblast</em>. As output, for each base, there is the adjusted p valued (FDR).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_value_wilcox_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_wilcox_test.html">get_wilcox_test</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">[</span>, <span class="va">relevant_bases</span><span class="op">]</span>, <span class="va">cluster_ci</span>, <span class="st">"Winner Epiblast"</span>, <span class="st">"Loser Epiblast"</span> , <span class="va">index_ci</span><span class="op">)</span></code></pre></div>
<p>We sort the bases according to the adjusted p value in order to identify the bases where the heteroplasmy is most different between the two clusters.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_value_wilcox_test_sort</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">p_value_wilcox_test</span>, decreasing <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>The heteroplasmy and the corresponding allele frequencies for the most relevant bases (according to Wilcoxon test) are shown. We can notice that the top 10 relevant bases belong to the genes mt-Rnr1 and mt-Rnr1 and in these positions the heteroplasmy is always higher in the <em>Loser Epiblast</em>. Below the customized functions <em>plot_heteroplasmy_epiblast</em> and <em>plot_allele_frequency_epiblast</em> are used in order to keep the same colour assignment and notation used in <a href="https://www.nature.com/articles/s42255-021-00422-7" class="external-link">Lima et al, 2021</a>. For general use, please use functions <em>plot_heteroplasmy</em> and <em>plot_allele_frequency</em></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">p_value_wilcox_test_sort</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_heteroplasmy_epiblast</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">p_value_wilcox_test_sort</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">cluster_ci</span>, <span class="va">index_ci</span><span class="op">)</span><span class="op">+</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">p_value_wilcox_test_sort</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_value_wilcox_test_sort</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="fl">4</span><span class="op">)</span>, sep <span class="op">=</span>  <span class="st">"-"</span><span class="op">)</span><span class="op">)</span>
<span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">q</span>, <span class="va">p</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">q</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; [[1]][[1]]</span>
<span class="co">#&gt; list()</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[1]][[2]]</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
<pre><code><span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span></code></pre>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-15-2.png" width="672"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">


<span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">p_value_wilcox_test_sort</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_allele_frequency_epiblast</span><span class="op">(</span><span class="va">i</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">allele_matrix_ci</span>, <span class="va">cluster_ci</span>, <span class="va">name_position_qc</span>, <span class="va">name_position_allele_qc</span>, <span class="fl">5</span>, <span class="va">index_ci</span><span class="op">)</span>  
<span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">q</span>, <span class="va">p</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-15-3.png" width="672"><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-15-4.png" width="672"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; [[1]][[1]]</span>
<span class="co">#&gt; list()</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[1]][[2]]</span>
<span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span>
<span class="co">#&gt;   z     cells    name           grob</span>
<span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span>
<span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span>
<span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span>
<span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; TableGrob (2 x 2) "arrange": 4 grobs</span>
<span class="co">#&gt;   z     cells    name           grob</span>
<span class="co">#&gt; 1 1 (1-1,1-1) arrange gtable[layout]</span>
<span class="co">#&gt; 2 2 (1-1,2-2) arrange gtable[layout]</span>
<span class="co">#&gt; 3 3 (2-2,1-1) arrange gtable[layout]</span>
<span class="co">#&gt; 4 4 (2-2,2-2) arrange gtable[layout]</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="identification-of-most-different-bases-according-to-heteroplasmy-along-pseudo-time">Identification of most different bases according to heteroplasmy along pseudo-time<a class="anchor" aria-label="anchor" href="#identification-of-most-different-bases-according-to-heteroplasmy-along-pseudo-time"></a>
</h3>
<p>If for each cell the diffusion pseudo-time information is also available, then it is possible to detect the bases whose heteroplasmy changes in a significant way along pseudo-time. To perform this task three different methods are available: <em>GAM</em> (the adjusted p value provided by the GAM fit is assigned to each base), <em>pearson</em> and <em>spearman</em> (the adjusted p value provided by the pearson or spearman correlation test between heteroplasmy and diffusion pseudo-time is assigned to each base). For more info see <em>?dpt_test</em></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">time</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">pseudo_time</span>

<span class="va">dpt_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dpt_test.html">dpt_test</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">[</span>, <span class="va">relevant_bases</span><span class="op">]</span>, <span class="va">time</span>, <span class="va">index_ci</span>, method <span class="op">=</span>  <span class="st">"GAM"</span><span class="op">)</span>

<span class="va">dpt_analysis_result_final</span> <span class="op">&lt;-</span> <span class="va">dpt_analysis</span><span class="op">[</span><span class="va">dpt_analysis</span><span class="op">$</span><span class="va">FDR_value</span><span class="op">&lt;</span><span class="fl">0.001</span>, <span class="op">]</span></code></pre></div>
<p>The heteroplasmy level for the top 2 relevant bases (according to GAM fit) is shown along pseudo time, with the black line denoting the GAM fit. We can notice that the top bases belong to the genes mt-Rnr1 and mt-Rnr2 and in these positions the heteroplasmy increases with the diffusion pseudo time. The top 10 positions identified with the diffusion pseudo time analysis are equal to the top 10 positions detected with the Wilcoxon test based on clustering partition. Below the customized function <em>plot_dpt_epiblast</em> is used in order to keep the same colour assignment and notation used in <a href="https://www.nature.com/articles/s42255-021-00422-7" class="external-link">Lima et al, 2021</a>. For general use, please use function <em>plot_dpt</em>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="va">dpt_analysis</span><span class="op">$</span><span class="va">Position</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_dpt_epiblast</span><span class="op">(</span><span class="va">i</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">cluster_ci</span>, <span class="va">time</span>, <span class="va">dpt_analysis</span>, <span class="va">index_ci</span><span class="op">)</span>
<span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">q</span>, <span class="va">p</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">q</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-options-for-top-bases-according-to-heteroplasmy">Plotting options for top bases according to heteroplasmy<a class="anchor" aria-label="anchor" href="#plotting-options-for-top-bases-according-to-heteroplasmy"></a>
</h3>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="va">bases_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"326_MT"</span>,<span class="st">"327_MT"</span>,<span class="st">"303_MT"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_correlation_bases.html">plot_correlation_bases</a></span><span class="op">(</span><span class="va">bases_example</span>, <span class="va">index_ci</span>, <span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">




<span class="fu"><a href="../reference/plot_spider_chart.html">plot_spider_chart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sum_matrix_qc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">cluster_ci</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">index_ci</span><span class="op">)</span>

<span class="fu"><a href="../reference/plot_spider_chart.html">plot_spider_chart</a></span><span class="op">(</span><span class="st">"326_MT"</span>, <span class="va">cluster_ci</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">index_ci</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_spider_chart.html">plot_spider_chart</a></span><span class="op">(</span><span class="st">"327_MT"</span>, <span class="va">cluster_ci</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">index_ci</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_spider_chart.html">plot_spider_chart</a></span><span class="op">(</span><span class="st">"303_MT"</span>, <span class="va">cluster_ci</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">index_ci</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R">




<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"coordinate_dm.Rda"</span>, package <span class="op">=</span> <span class="st">"MitoHEAR"</span><span class="op">)</span><span class="op">)</span>



<span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">cluster</span>

<span class="va">coordinate_dm</span> <span class="op">&lt;-</span> <span class="va">coordinate_dm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>, <span class="op">]</span>

<span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_coordinate_cluster_epiblast</span><span class="op">(</span><span class="va">coordinate_dm</span> , <span class="va">cluster</span><span class="op">)</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/plot_coordinate_heteroplasmy.html">plot_coordinate_heteroplasmy</a></span><span class="op">(</span><span class="va">coordinate_dm</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">index_ci</span>, <span class="st">"327_MT"</span><span class="op">)</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-20-2.png" width="672"></p>
<p>We can check if there is a batch effect in the most relevant positions (i.e. the heteroplasmy levels are constantly higher only in a specific batch). For the top 2 positions defined with the GAM fit there is not a batch effect, since the heteroplasmy levels are higher among all cells in the <em>Loser Epiblast</em> clusters and not only in a specific batch. Below the customized function <em>plot_batch_epiblast</em> is used in order to keep the same colour assignment and notation used in <a href="https://www.nature.com/articles/s42255-021-00422-7" class="external-link">Lima et al, 2021</a>. For general use please use function <em>plot_batch</em>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">after_qc_qc</span> <span class="op">&lt;-</span> <span class="va">after_qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"a)Win 1"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"a)Win 5"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"a)Win 4"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"a)Win 2"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"a)Win 3"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"b)Int 1"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"b)Int 5"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"b)Int 4"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"b)Int 2"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">3</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"b)Int 3"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"c)Los 1"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"5"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"c)Los 5"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"c)Los 4"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"c)Los 2"</span>
<span class="va">batch</span><span class="op">[</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">4</span><span class="op">&amp;</span><span class="va">after_qc_qc</span><span class="op">$</span><span class="va">batch</span> <span class="op">==</span> <span class="st">"3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"c)Los 3"</span>






<span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_batch_epiblast</span><span class="op">(</span><span class="va">dpt_analysis</span><span class="op">$</span><span class="va">Position</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">batch</span>, <span class="va">cluster_ci</span>, <span class="fl">6</span>, <span class="va">index_ci</span><span class="op">)</span>

<span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_batch_epiblast</span><span class="op">(</span><span class="va">dpt_analysis</span><span class="op">$</span><span class="va">Position</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">batch</span>, <span class="va">cluster_ci</span>, <span class="fl">6</span>, <span class="va">index_ci</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="unsupervised-cluster-analysis-among-cells-based-on-allele-frequency-values">Unsupervised cluster analysis among cells based on allele frequency values<a class="anchor" aria-label="anchor" href="#unsupervised-cluster-analysis-among-cells-based-on-allele-frequency-values"></a>
</h3>
<p>MitoHEAR offers the possibility to perform an unsupervised hierarchical clustering on the cells based on a distance matrix with the function <em>clustering_dist_ang</em>. Given a base, the distance between two cells is the angular distance of the allele frequencies. Given a base, the variance of the distance values between two cells is also computed. Top bases with highest variance are selected for down stream analysis. We can represent the difference between two cells as a vector whose coordinates are the angular distances of the top bases. The total distance between two cell is the euclidean norm of the vector of difference between the two cells. The output of <em>clustering_dist_ang</em> is a list. The first element is a data frame which contains the old classification (partition available before the cluster analysis based on allele frequencies) and the new classification (partition provided by the cluster analysis based on allele frequencies ). The second element is the distance matrix, on which the hierarchical clustering is done. The third element is a vector with the top bases according to variance.</p>
<p>It is also possible to run <em>clustering_dist_ang</em> in a supervised approach. In this case the bases used for hierarchical clustering are not selected according to variance, but are directly provided with the parameter <em>relevant_bases</em>. The heatmap of the distance matrix with cells sorted according to the new classification is shown below. The cluster analysis based on allele frequencies information can be a powerful way to perform a lineage tracing analysis, by grouping together cells which are from the same embryo. See <em>?clustering_dist_ang</em> for more info.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result_clustering_sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustering_angular_distance.html">clustering_angular_distance</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span>, <span class="va">allele_matrix_ci</span>, <span class="va">cluster_ci</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">heteroplasmy_matrix_ci</span><span class="op">)</span><span class="op">)</span>, deepSplit_param <span class="op">=</span> <span class="fl">0</span>, minClusterSize_param <span class="op">=</span> <span class="fl">50</span>, <span class="fl">0.2</span>, min_value <span class="op">=</span> <span class="fl">0.001</span>, index <span class="op">=</span> <span class="va">index_ci</span>, relevant_bases <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="co">#&gt;  ..cutHeight not given, setting it to 0.289  ===&gt;  99% of the (truncated) height range in dendro.</span>
<span class="co">#&gt;  ..done.</span>


<span class="va">old_new_classification</span> <span class="op">&lt;-</span> <span class="va">result_clustering_sc</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">dist_matrix_sc</span> <span class="op">&lt;-</span> <span class="va">result_clustering_sc</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">top_dist</span> <span class="op">&lt;-</span> <span class="va">result_clustering_sc</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="va">common_idx</span> <span class="op">&lt;-</span> <span class="va">result_clustering_sc</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>

<span class="va">old_classification</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">old_new_classification</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">new_classification</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">old_new_classification</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>

<span class="fu"><a href="../reference/plot_distance_matrix.html">plot_distance_matrix</a></span><span class="op">(</span><span class="va">dist_matrix_sc</span>, <span class="va">old_classification</span><span class="op">)</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<p>Below the top 2 bases selected for the unsupervised cluster analysis. Consistently these are the same top 2 bases according to GAM fit.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">q</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">top_dist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="va">p</span> <span class="op">&lt;-</span>  <span class="fu">MitoHEAR</span><span class="fu">:::</span><span class="fu">plot_heteroplasmy_epiblast</span><span class="op">(</span><span class="va">top_dist</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">heteroplasmy_matrix_ci</span>, <span class="va">cluster_ci</span>, <span class="va">index_ci</span><span class="op">)</span>
<span class="va">q</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">q</span>, <span class="va">p</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">q</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; [[1]][[1]]</span>
<span class="co">#&gt; list()</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[1]][[2]]</span></code></pre></div>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<pre><code><span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span></code></pre>
<p><img src="cell_competition_mt_example_notebook_files/figure-html/unnamed-chunk-23-2.png" width="672"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Mojave 10.14.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] MitoHEAR_0.1.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] backports_1.4.1             circlize_0.4.13            </span>
<span class="co">#&gt;   [3] Hmisc_4.6-0                 BiocFileCache_1.14.0       </span>
<span class="co">#&gt;   [5] systemfonts_1.0.4           lazyeval_0.2.2             </span>
<span class="co">#&gt;   [7] splines_4.0.2               BiocParallel_1.24.1        </span>
<span class="co">#&gt;   [9] GenomeInfoDb_1.26.7         ggplot2_3.3.5              </span>
<span class="co">#&gt;  [11] digest_0.6.29               foreach_1.5.1              </span>
<span class="co">#&gt;  [13] ensembldb_2.14.1            htmltools_0.5.2            </span>
<span class="co">#&gt;  [15] magick_2.7.3                fansi_0.5.0                </span>
<span class="co">#&gt;  [17] magrittr_2.0.1              checkmate_2.0.0            </span>
<span class="co">#&gt;  [19] memoise_2.0.1               BSgenome_1.58.0            </span>
<span class="co">#&gt;  [21] cluster_2.1.0               ComplexHeatmap_2.6.2       </span>
<span class="co">#&gt;  [23] Biostrings_2.58.0           rdist_0.0.5                </span>
<span class="co">#&gt;  [25] matrixStats_0.61.0          askpass_1.1                </span>
<span class="co">#&gt;  [27] pkgdown_2.0.2               prettyunits_1.1.1          </span>
<span class="co">#&gt;  [29] jpeg_0.1-9                  colorspace_2.0-2           </span>
<span class="co">#&gt;  [31] blob_1.2.2                  rappdirs_0.3.3             </span>
<span class="co">#&gt;  [33] textshaping_0.3.6           xfun_0.29                  </span>
<span class="co">#&gt;  [35] dplyr_1.0.7                 crayon_1.4.2               </span>
<span class="co">#&gt;  [37] RCurl_1.98-1.5              jsonlite_1.7.2             </span>
<span class="co">#&gt;  [39] survival_3.2-3              VariantAnnotation_1.36.0   </span>
<span class="co">#&gt;  [41] iterators_1.0.13            glue_1.6.0                 </span>
<span class="co">#&gt;  [43] gtable_0.3.0                zlibbioc_1.36.0            </span>
<span class="co">#&gt;  [45] XVector_0.30.0              GetoptLong_1.0.5           </span>
<span class="co">#&gt;  [47] DelayedArray_0.16.3         shape_1.4.6                </span>
<span class="co">#&gt;  [49] BiocGenerics_0.36.1         scales_1.1.1               </span>
<span class="co">#&gt;  [51] bezier_1.1.2                DBI_1.1.2                  </span>
<span class="co">#&gt;  [53] Rcpp_1.0.7                  fmsb_0.7.2                 </span>
<span class="co">#&gt;  [55] progress_1.2.2              htmlTable_2.4.0            </span>
<span class="co">#&gt;  [57] clue_0.3-60                 foreign_0.8-80             </span>
<span class="co">#&gt;  [59] bit_4.0.4                   Formula_1.2-4              </span>
<span class="co">#&gt;  [61] stats4_4.0.2                karyoploteR_1.16.0         </span>
<span class="co">#&gt;  [63] htmlwidgets_1.5.4           httr_1.4.2                 </span>
<span class="co">#&gt;  [65] RColorBrewer_1.1-2          ellipsis_0.3.2             </span>
<span class="co">#&gt;  [67] pkgconfig_2.0.3             XML_3.99-0.8               </span>
<span class="co">#&gt;  [69] farver_2.1.0                nnet_7.3-14                </span>
<span class="co">#&gt;  [71] sass_0.4.0                  dbplyr_2.1.1               </span>
<span class="co">#&gt;  [73] dynamicTreeCut_1.63-1       utf8_1.2.2                 </span>
<span class="co">#&gt;  [75] tidyselect_1.1.1            labeling_0.4.2             </span>
<span class="co">#&gt;  [77] rlang_0.4.12                AnnotationDbi_1.52.0       </span>
<span class="co">#&gt;  [79] munsell_0.5.0               tools_4.0.2                </span>
<span class="co">#&gt;  [81] cachem_1.0.6                generics_0.1.1             </span>
<span class="co">#&gt;  [83] RSQLite_2.2.9               evaluate_0.14              </span>
<span class="co">#&gt;  [85] stringr_1.4.0               fastmap_1.1.0              </span>
<span class="co">#&gt;  [87] yaml_2.2.1                  ragg_1.2.2                 </span>
<span class="co">#&gt;  [89] knitr_1.37                  bit64_4.0.5                </span>
<span class="co">#&gt;  [91] fs_1.5.2                    purrr_0.3.4                </span>
<span class="co">#&gt;  [93] AnnotationFilter_1.14.0     gam_1.20                   </span>
<span class="co">#&gt;  [95] xml2_1.3.3                  biomaRt_2.46.3             </span>
<span class="co">#&gt;  [97] compiler_4.0.2              rstudioapi_0.13            </span>
<span class="co">#&gt;  [99] curl_4.3.2                  png_0.1-7                  </span>
<span class="co">#&gt; [101] tibble_3.1.6                bslib_0.3.1                </span>
<span class="co">#&gt; [103] stringi_1.7.6               highr_0.9                  </span>
<span class="co">#&gt; [105] GenomicFeatures_1.42.3      desc_1.2.0                 </span>
<span class="co">#&gt; [107] lattice_0.20-41             ProtGenerics_1.22.0        </span>
<span class="co">#&gt; [109] Matrix_1.3-4                vctrs_0.3.8                </span>
<span class="co">#&gt; [111] pillar_1.6.4                lifecycle_1.0.1            </span>
<span class="co">#&gt; [113] jquerylib_0.1.4             GlobalOptions_0.1.2        </span>
<span class="co">#&gt; [115] data.table_1.14.2           bitops_1.0-7               </span>
<span class="co">#&gt; [117] rtracklayer_1.50.0          GenomicRanges_1.42.0       </span>
<span class="co">#&gt; [119] R6_2.5.1                    latticeExtra_0.6-29        </span>
<span class="co">#&gt; [121] gridExtra_2.3               IRanges_2.24.1             </span>
<span class="co">#&gt; [123] codetools_0.2-16            dichromat_2.0-0            </span>
<span class="co">#&gt; [125] assertthat_0.2.1            SummarizedExperiment_1.20.0</span>
<span class="co">#&gt; [127] openssl_1.4.6               rprojroot_2.0.2            </span>
<span class="co">#&gt; [129] rjson_0.2.20                regioneR_1.22.0            </span>
<span class="co">#&gt; [131] GenomicAlignments_1.26.0    Rsamtools_2.6.0            </span>
<span class="co">#&gt; [133] S4Vectors_0.28.1            GenomeInfoDbData_1.2.4     </span>
<span class="co">#&gt; [135] parallel_4.0.2              hms_1.1.1                  </span>
<span class="co">#&gt; [137] grid_4.0.2                  rpart_4.1-15               </span>
<span class="co">#&gt; [139] bamsignals_1.22.0           rmarkdown_2.11             </span>
<span class="co">#&gt; [141] MatrixGenerics_1.2.1        Cairo_1.5-12.2             </span>
<span class="co">#&gt; [143] biovizBase_1.38.0           Biobase_2.50.0             </span>
<span class="co">#&gt; [145] base64enc_0.1-3</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Gabriele Lubatti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
