<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>get_raw_counts_allele — get_raw_counts_allele • MitoHEAR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="get_raw_counts_allele — get_raw_counts_allele"><meta property="og:description" content="It is one the two main function of the MitoHEAR package (together
with get_heteroplasmy). The function allows to obtain a matrix of
counts (n_row = number of sample, n_col= 4*number of bases) of the four
alleles in each base, for every sample. It takes as input a vector of sorted
bam files (one bam file for each sample) and a fasta file for the genomic
region of interest. It is based on the pileup function of the package
Rsamtools."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MitoHEAR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Ludwig_et_al_example_notebook.html">Lineage tracing with mtDNA heteroplasmy in single cell RNA seq human data</a>
    </li>
    <li>
      <a href="../articles/cell_competition_bulk_data_mt_example_notebook.html">Analysing mtDNA cell lines in bulk RNA seq data</a>
    </li>
    <li>
      <a href="../articles/cell_competition_mt_example_notebook.html">Analysing mtDNA heteroplasmy in single cell RNA seq mouse data</a>
    </li>
    <li>
      <a href="../articles/lineage_tracing_example_notebook.html">Lineage tracing with mtDNA heteroplasmy in single cell RNA seq mouse data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ScialdoneLab/MitoHEAR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>get_raw_counts_allele</h1>
    <small class="dont-index">Source: <a href="https://github.com/ScialdoneLab/MitoHEAR/blob/HEAD/R/get_raw_counts_allele.R" class="external-link"><code>R/get_raw_counts_allele.R</code></a></small>
    <div class="hidden name"><code>get_raw_counts_allele.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>It is one the two main function of the <strong>MitoHEAR</strong> package (together
with <em>get_heteroplasmy</em>). The function allows to obtain a matrix of
counts (n_row = number of sample, n_col= 4*number of bases) of the four
alleles in each base, for every sample. It takes as input a vector of sorted
bam files (one bam file for each sample) and a fasta file for the genomic
region of interest. It is based on the <em>pileup</em> function of the package
Rsamtools.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_raw_counts_allele</span><span class="op">(</span><span class="va">bam_input</span>, <span class="va">path_fasta</span>, <span class="va">cell_names</span>, cores_number <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bam_input</dt>
<dd><p>Character vector of sorted bam files (full path). Each
sample is defined by one bam file. For each bam file it is needed also the
index bam file (.bai) at the same path.</p></dd>
<dt>path_fasta</dt>
<dd><p>Character string with full path to the fasta file of the
genomic region of interest.</p></dd>
<dt>cell_names</dt>
<dd><p>Character vector of sample names.</p></dd>
<dt>cores_number</dt>
<dd><p>Number of cores to use.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list with three elements:</p>
<dl><dt>matrix_allele_counts </dt>
<dd><p>Matrix of counts (n_row = number of sample,
n_col= 4*number of bases) of the four alleles in each base, for every
sample. The row names is equal to cell_names.</p></dd>
<dt>name_position_allele</dt>
<dd><p>Character vector with length equal to n_col of
matrix_allele_counts. Each element specifies the coordinate of genomic
position for a base and the allele.</p></dd>
<dt>name_position</dt>
<dd><p>Character vector
with length equal to n_col of matrix_allele_counts. Each element specifies
the coordinate of genomic position for a base.</p></dd>
</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="https://www.rdocumentation.org/packages/Rsamtools/versions/1.24.0/topics/pileup" class="external-link">https://www.rdocumentation.org/packages/Rsamtools/versions/1.24.0/topics/pileup</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Gabriele Lubatti <a href="mailto:gabriele.lubatti@helmholtz-muenchen.de">gabriele.lubatti@helmholtz-muenchen.de</a></p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gabriele Lubatti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

